cmds/receive.c

```diff
@@ -643,18 +647,24 @@ static int process_rmdir(const char *path, void *user)
 	return ret;
 }
 
-static int open_inode_for_write(struct btrfs_receive *rctx, const char *path)
+static int open_inode_for_write(struct btrfs_receive *rctx, const char *path,
+				int extra_flags)
 {
 	int ret = 0;
 
 	if (rctx->write_fd != -1) {
-		if (strcmp(rctx->write_path, path) == 0)
+		/*
+		 * if the existing fd is for this path and the needed flags are
+		 * satisfied, no need to open a new one
+		 */
+		if (strcmp(rctx->write_path, path) == 0 &&
+		    (!extra_flags || rctx->write_fd_flags & extra_flags))
```

Comment:
Good point.. I could change it to something like (extra_flags & rctx->write_fd_flags == extra_flags). I guess it doesn't feel quite right to have special code for O_ALLOW_ENCODED, but I can do it that way since I'm clearly repeatedly screwing up doing it the more generic way, so your point is taken :)

Reply:
how do you feel about redundant calls to fcntl if the bool is passed in?

Reply:
I'd track the state so you don't need to do unnecessary calls. Something along the lines of this:

```c
if (strcmp(rctx->write_path, path) == 0) {
	if (allow_encoded && !rctx->write_fd_allow_encoded) {
		fcntl thing
		rctx->write_fd_allow_encoded = true;
	}
	goto out;
}
flags = O_RDWR;
if (allow_encoded)
	flags |= O_ALLOW_ENCODED;
rctx->write_fd = open(path, flags);
...
rctx->write_fd_allow_encoded = allow_encoded;
```

Reply:
right on, thanks.